src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlRetry/SqlRetryService.cs

```diff
@@ -0,0 +1,247 @@
+
+            _defaultIsExceptionRetriableOff = sqlRetryServiceDelegateOptions.DefaultIsExceptionRetriableOff;
+            _customIsExceptionRetriable = sqlRetryServiceDelegateOptions.CustomIsExceptionRetriable;
+        }
+
+        public delegate bool IsExceptionRetriable(Exception ex);
+
+        private static bool DefaultIsExceptionRetriable(Exception ex)
+        {
+            if (ex is SqlException sqlEx)
+            {
+                if (sqlEx.Number == 121 && sqlEx.Message.Contains("an error occurred during the pre-login handshake", StringComparison.OrdinalIgnoreCase))
+                {
+                    return true;
+                }
+
+                if (sqlEx.ToString().Contains("login failed", StringComparison.OrdinalIgnoreCase))
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        private bool RetryTest(Exception ex)
+        {
+            if (ex is SqlException sqlEx && _transientErrors.Contains(sqlEx.Number))
+            {
+                return true;
+            }
+
+            if (!_defaultIsExceptionRetriableOff && _defaultIsExceptionRetriable(ex))
+            {
+                return true;
+            }
+
+            if (_customIsExceptionRetriable?.Invoke(ex) == true)
+            {
+                return true;
+            }
+
+            if (ex.IsRetriable())
+            {
+                return true;
+            }
+
+            return false;
+        }
+
+        public async Task ExecuteSql(Func<CancellationToken, Task> action, CancellationToken cancellationToken)
+        {
+            EnsureArg.IsNotNull(action, nameof(action));
+
+            int retry = 0;
+            while (true)
+            {
+                try
+                {
+                    await action(cancellationToken);
+                    return;
+                }
+                catch (Exception ex)
+                {
+                    if (!RetryTest(ex) || ++retry >= _maxRetries)
+                    {
+                        throw;
+                    }
+                }
+
+                await Task.Delay(_retryMillisecondsDelay, cancellationToken);
+            }
+        }
+
+        public async Task ExecuteSql<TLogger>(SqlCommand sqlCommand, Func<SqlCommand, CancellationToken, Task> action, ILogger<TLogger> logger, string logMessage, CancellationToken cancellationToken)
+        {
+            EnsureArg.IsNotNull(sqlCommand, nameof(sqlCommand));
+            EnsureArg.IsNotNull(action, nameof(action));
+            EnsureArg.IsNotNull(logger, nameof(logger));
+            EnsureArg.IsNotNull(logMessage, nameof(logMessage));
+
+            int retry = 0;
+            while (true)
+            {
+                try
+                {
+                    // WARNING, this code will not set sqlCommand.Transaction. Sql transactions via C#/.NET are not supported in this method.
```

Comment:
Should be the method documentation

Reply:
Added it to the class doc since it applies to all the methods.